Add code to automatically disable inactive FreeIPA users

jsf9k opened the pull request:
## 🗣 Description ##

This pull request:
- Adds code to disable the `KDC:Disable Last Success` password plugin feature.
- Replaces the old script to disable inactive users with a new one that actually works.
- Creates a systemd timer and service to run the script regularly.

See also cisagov/freeipa-server-packer#133.

## 💭 Motivation and context ##

Resolves cisagov/cool-system-internal#143.

- We don't want the `KDC:Disable Last Success` password plugin feature enabled because we want to be able to disable inactive users, which requires us to be able to determine the last time a user authenticated.
- The old script didn't work.
- We need to disable inactive users at a regular cadence.

## 🧪 Testing ##

All automated tests pass.  I also built and tested a staging AMI with these changes as part of cisagov/freeipa-server-packer#133.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [x] Make `build / test (disable_inactive_users)` GitHub Actions checks required.

jsf9k committed:
Add code to disable KDC:Disable Last Success password plugin feature

We don't want the KDC:Disable Last Success password plugin feature
enabled because we want to be able to disable inactive users, which
requires us to be able to determine the last time a user
authenticated.

jsf9k committed:
Add a new and improved script to disable inactive users

The old script did not actually work as intended.

jsf9k committed:
Add a new role variable to determine if inactive users will be disabled

Also add a new scenario to test the case where inactive users are to
be disabled.

jsf9k committed:
Move trace and HSTS disabling code to the end of the file

This is a bit cleaner and makes the file less cluttered.

jsf9k committed:
Create a block to group the DHS CA tasks

jsf9k committed:
Convert the cron job to a systemd timer and service

jsf9k committed:
Rename resources to include "freeipa" for clarity

Given that these resources aren't located in FreeIPA-specific
locations it makes sense to include "freeipa" in the name.

jsf9k committed:
Remove code to install cronie on Fedora

We no longer use cron functionality in this role.

jsf9k committed:
Add RandomizedDelaySec parameter

We don't want two FreeIPA servers to disable inactive users at the
same time, since there is a small chance they may collide.

jsf9k committed:
Uncomment command that actually disables the user

jsf9k committed:
Add to FreeIPA servers the ability to enable and disable users

jsf9k committed:
Use the FreeIPA server host's keytab to kinit

This ensures we have the necessary permissions to enable and disable
users.

jsf9k commented on files/disable-inactive-freeipa-users.service:
```diff
@@ -0,0 +1,7 @@
+Description=Disable inactive FreeIPA users.
+
+[Service]
+ExecStart=/usr/local/sbin/disable_inactive_freeipa_users.sh
```
This file should be converted to a template, with the change:
```suggestion
ExecStart=/usr/local/sbin/disable_inactive_freeipa_users.sh "{{ freeipa_days_before_inactive }}"
```

jsf9k replied:
See commit [c07572d](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/c07572dc31b7825c298b813d5de7e5edb47afcca).

jsf9k commented on files/disable-inactive-freeipa-users.service:
```diff
@@ -0,0 +1,7 @@
+
+[Service]
+ExecStart=/usr/local/sbin/disable_inactive_freeipa_users.sh
+PrivateTmp=true
```
There are [other sandboxing-related directives](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html) that could easily be added here, but we'd have to re-test.  Create an issue so we do not forget to circle back and lock down this service as much as possible.

jsf9k replied:
See commit 5aabe0b.

jsf9k committed:
Convert the systemd service file to a template

This allows us to make use of the freeipa_days_before_inactive role
variable to specify the number of days before an unauthenticated user
is considered inactive.

jsf9k committed:
Remove unnecessary test

This test was erroneously left over from a previous copy and paste.

jsf9k committed:
Add a comment reminding us to add all relevant sandboxing directives

jsf9k committed:
Start the systemd timer

Just enabling the timer doesn't do anything until we reboot the
system.

dav3r commented on defaults/main.yml:
```diff
@@ -3,6 +3,10 @@
 freeipa_days_before_inactive: 45
 
+# Whether or not to set up FreeIPA and create a cron job to disable
+# inactive users.
```
Since we are not using `cron` anymore.  Feel free to reword this comment further if you so choose:
```suggestion
# Whether or not to set up FreeIPA and create a systemd timer and
# service to disable inactive users.
```

dav3r commented on templates/disable-inactive-freeipa-users.j2.service:
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Disable inactive FreeIPA users.
+
+[Service]
```
Do we want to alphabetize these `[Unit]` and `[Service]` blocks?  I'm fine with it as is, just figured I'd stoke the conversation here.

jsf9k replied:
The current order is [the standard for `systemd` services](https://man.archlinux.org/man/systemd.service.5#EXAMPLES).  `systemd` timers, services, sockets, etc. are all units, and a service is a particular type of unit.  So the sections are in order from most general to most specific.

dav3r replied:
Makes sense, thanks for the info. 👍

dav3r commented on files/01_setup_disabling_of_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,70 @@
+set -o errexit
+set -o pipefail
+
+# Set up this instance to disable inactive users.
```
Just to be extra clear about which users we are referring to.
```suggestion
# Set up this instance to disable inactive FreeIPA users.
```

dav3r commented on files/01_setup_disabling_of_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,70 @@
+  done
+  # Run the ipa config-mod command.
+  #
+  # Note that it is harmless to run this command when it changes
+  # nothing; but, we must temporarily turn off the bash option errexit
+  # since in that case the error code indicates a failure.
```
Great comment, thanks! 👍

dav3r commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,130 @@
+# Usage:
+#   disable_inactive_freeipa_users.sh [disable_time]
+#
+# A user is considered inactive if he or she has not logged since
```
Use more inclusive wording and add a missing word:
```suggestion
# A user is considered inactive if they have not logged in since
```

jsf9k replied:
"They" is a plural pronoun.  How about this:
```suggestion
# Users are considered inactive if they have not logged in since
```

jsf9k replied:
See [8d971f1](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/8d971f1de83ce7394cb916e1fc1e2ce418671208) for the missing word.

dav3r replied:
> "They" is a plural pronoun. How about this:

That works for me. 👍

jsf9k replied:
Please see commit [43c6240](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/43c6240629a59d716428f7602f86189967251f81).

dav3r commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,130 @@
+# disable_time.  If disable_time is not specified then a default value
+# of "45 days ago" is used.
+#
+# The date/time must be specified so that it understandable by the
```
Add missing word:
```suggestion
# The date/time must be specified so that it is understandable by the
```

dav3r commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,130 @@
+set -o errexit
+set -o pipefail
+
+# A user is considered inactive if he or she has not logged since this
```
Use more inclusive wording and add a missing word:
```suggestion
# A user is considered inactive if they have not logged in since this
```

jsf9k replied:
See [8d971f1](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/8d971f1de83ce7394cb916e1fc1e2ce418671208) for the missing word.

jsf9k replied:
Please see commit [43c6240](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/43c6240629a59d716428f7602f86189967251f81).

dav3r commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,130 @@
+disable_time=${1:-45 days ago}
+
+# Use a temp file for our Kerberos credentials cache, since it only
+# needs to exit while this script is running.
```
I assume this is what you meant?
```suggestion
# needs to exist while this script is running.
```

dav3r commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,130 @@
+Usage:
+  ${0##*/} [disable_time]
+
+A user is considered inactive if he or she has not logged since
```
Use more inclusive wording and add a missing word:
```suggestion
A user is considered inactive if they have not logged in since
```

jsf9k replied:
See [8d971f1](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/8d971f1de83ce7394cb916e1fc1e2ce418671208) for the missing word.

jsf9k replied:
Please see commit [43c6240](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/43c6240629a59d716428f7602f86189967251f81).

dav3r commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,130 @@
+disable_time.  If disable_time is not specified then a default value
+of "45 days ago" is used.
+
+The date/time must be specified so that it understandable by the date
```
Add missing word:
```suggestion
The date/time must be specified so that it is understandable by the date
```

dav3r commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,130 @@
+    if $disable_user; then
+      if ! $all_timestamps_invalid; then
+        ipa user-disable "$user"
+        echo User "$user" disabled because he or she is an inactive user.
+      else
+        echo User "$user" not disabled because all authentication timestamps were invalid.
+      fi
+    else
+      echo User "$user" not disabled because he or she is an active user.
```
Use more inclusive wording:
```suggestion
        echo User "$user" disabled because they are an inactive user.
      else
        echo User "$user" not disabled because all authentication timestamps were invalid.
      fi
    else
      echo User "$user" not disabled because they are an active user.
```

jsf9k replied:
"They" is a plural pronoun.  How about:
```suggestion
        echo User "$user" disabled due to inactivity.
      else
        echo User "$user" not disabled because all authentication timestamps were invalid.
      fi
    else
      echo User "$user" not disabled due to sufficiently recent authentication.
```

dav3r replied:
That works for me.

jsf9k replied:
Please see commit [43c6240](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/43c6240629a59d716428f7602f86189967251f81).

dav3r commented on molecule/disable_inactive_users/tests/test_disable_inactive_users.py:
```diff
@@ -0,0 +1,42 @@
+"""Module containing the tests for the disable_trace scenario."""
```
Reference the correct scenario name:
```suggestion
"""Module containing the tests for the disable_inactive_users scenario."""
```

dav3r reviewed (changes requested):
I didn't notice anything major, but I did note a bunch of small thangs to clean up.

jsf9k committed:
Be clear as to which type of users we are referring to

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Replace reference to cron with mention of systemd service and timer

We no longer use cron.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Fix assorted typos and copy/paste errors

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Add missing word

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>
Suggested-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

jsf9k committed:
Update language to be more inclusive

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>
Suggested-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

dav3r reviewed (approved):
Looking great, strong work! 🦾 💼

jsf9k committed:
Fix two places where I wrote "delete" but meant "disable"

Co-authored-by: Dave Redmin <david.redmin@trio.dhs.gov>
Reported-by: Dave Redmin <david.redmin@trio.dhs.gov>

jsf9k commented:
@dav3r - Re our conversation yesterday, please see commit [2d3069d](https://github.com/cisagov/ansible-role-freeipa-server/pull/73/commits/2d3069d36a5cf56da2353743347749a7f9a7b1f2).

jsf9k committed:
Fix typo

+%S returns the number of seconds into the current minute, whereas +%s
returns the number of seconds since the epoch.

When +%S returned a number with a leading zero the test ((timestamp >
disable_deadline)) would fail and the user would get erroneously
disabled.

jsf9k committed:
Delete the temporary KRB credentials cache file

There is no sense in letting these files take up space in the /tmp
directory.

dav3r reviewed (approved):
Even better the second time! 🚀

mcdonnnj commented on README.md:
```diff
@@ -15,6 +15,7 @@ None.
 | Variable | Description | Default | Required |
 |----------|-------------|---------|----------|
 | freeipa_days_before_inactive | The number of days a user can go without logging in before his or her account is determined to be inactive and is disabled. | `45` | No |
+| freeipa_disable_inactive_users | Whether or not to set up FreeIPA and systemd to periodically disable inactive users. | `false` | No |
```
Just to be consistent I think this should match the comment in `defaults/main.yml`.

```suggestion
| freeipa_disable_inactive_users | Whether or not to set up FreeIPA and create a systemd timer and service to disable inactive users. | `false` | No |
```

mcdonnnj commented on files/01_setup_disabling_of_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,70 @@
+  # disable inactive users, which requires us to be able to determine
+  # the last time a user authenticated.
+  password_plugin_features=$(ipa config-show \
+    |
+    # --quiet means no printing unless the p command is used
+    sed --quiet "s/^\s*Password plugin features:\s*\(.*\)/\1/p" \
```
Is there a reason it is not like this?

```suggestion
    # --quiet means no printing unless the p command is used
    | sed --quiet "s/^\s*Password plugin features:\s*\(.*\)/\1/p" \
```

mcdonnnj commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,131 @@
+    # Note that we disable the size and time limits normally in place
+    # when running ipa user-find.
+    ipa user-find --disabled=false --sizelimit=0 --timelimit=0 \
+      |
+      # --quiet means no printing unless the p command is used
+      sed --quiet 's/^\s*User login:\s*//p'
```
Same question as the previous.

```suggestion
      # --quiet means no printing unless the p command is used
      | sed --quiet 's/^\s*User login:\s*//p'
```

mcdonnnj commented on files/disable_inactive_freeipa_users.sh:
```diff
@@ -0,0 +1,131 @@
+
+  for user in $users; do
+    timestamps=$(ipa user-status "$user" \
+      |
+      # --quiet means no printing unless the p command is used
+      sed --quiet 's/\s*Last successful authentication:\s*//p')
```
One more.

```suggestion
      # --quiet means no printing unless the p command is used
      | sed --quiet 's/\s*Last successful authentication:\s*//p')
```

mcdonnnj commented on README.md:
```diff
@@ -15,6 +15,7 @@ None.
 | Variable | Description | Default | Required |
 |----------|-------------|---------|----------|
 | freeipa_days_before_inactive | The number of days a user can go without logging in before his or her account is determined to be inactive and is disabled. | `45` | No |
```
Inclusivee language change. The comment in `defaults/main.yml` should also be updated.

```suggestion
| freeipa_days_before_inactive | The number of days a user can go without logging in before their account is determined to be inactive and is disabled. | `45` | No |
```

jsf9k replied:
Please see commit 206a290.

jsf9k committed:
Update role var description to match comment in defaults/main.yml

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Rework wording to remove unintentionally non-inclusive language

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented:
@mcdonnnj - Your suggested changes regarding the location of pipe characters resulted in the following from the `shellcheck` linter:
```console
shellcheck...............................................................Failed
- hook id: shellcheck
- exit code: 1

In files/01_setup_disabling_of_inactive_freeipa_users.sh line 13:
  password_plugin_features=$(ipa config-show \
  ^-- SC1009 (info): The mentioned syntax error was in this variable assignment.
                           ^-- SC1073 (error): Couldn't parse this command expansion. Fix to allow more checks.


In files/01_setup_disabling_of_inactive_freeipa_users.sh line 15:
    | sed --quiet "s/^\s*Password plugin features:\s*\(.*\)/\1/p" \
    ^-- SC1072 (error): Expected end of $(..) expression. Fix any mentioned problems and try again.
    ^-- SC1133 (error): Unexpected start of line. If breaking lines, |/||/&& should be at the end of the previous one.


In files/disable_inactive_freeipa_users.sh line [63](https://github.com/cisagov/ansible-role-freeipa-server/actions/runs/10306946528/job/28535130174?pr=73#step:21:64):
  users=$(
  ^-- SC1009 (info): The mentioned syntax error was in this variable assignment.
        ^-- SC1073 (error): Couldn't parse this command expansion. Fix to allow more checks.


In files/disable_inactive_freeipa_users.sh line [68](https://github.com/cisagov/ansible-role-freeipa-server/actions/runs/10306946528/job/28535130174?pr=73#step:21:69):
      | sed --quiet 's/^\s*User login:\s*//p'
      ^-- SC1072 (error): Expected end of $(..) expression. Fix any mentioned problems and try again.
      ^-- SC1133 (error): Unexpected start of line. If breaking lines, |/||/&& should be at the end of the previous one.

For more information:
  https://www.shellcheck.net/wiki/SC1133 -- Unexpected start of line. If brea...
  https://www.shellcheck.net/wiki/SC10[72](https://github.com/cisagov/ansible-role-freeipa-server/actions/runs/10306946528/job/28535130174?pr=73#step:21:73) -- Expected end of $(..) expression....
  https://www.shellcheck.net/wiki/SC10[73](https://github.com/cisagov/ansible-role-freeipa-server/actions/runs/10306946528/job/28535130174?pr=73#step:21:74) -- Couldn't parse this command expan...
```

As a result I have dropped that commit.

mcdonnnj commented:
> @mcdonnnj - Your suggested changes regarding the location of pipe characters resulted in the following from the `shellcheck` linter:
> 
> ```
> shellcheck...............................................................Failed
> - hook id: shellcheck
> - exit code: 1
> 
> In files/01_setup_disabling_of_inactive_freeipa_users.sh line 13:
>   password_plugin_features=$(ipa config-show \
>   ^-- SC1009 (info): The mentioned syntax error was in this variable assignment.
>                            ^-- SC1073 (error): Couldn't parse this command expansion. Fix to allow more checks.
> 
> 
> In files/01_setup_disabling_of_inactive_freeipa_users.sh line 15:
>     | sed --quiet "s/^\s*Password plugin features:\s*\(.*\)/\1/p" \
>     ^-- SC1072 (error): Expected end of $(..) expression. Fix any mentioned problems and try again.
>     ^-- SC1133 (error): Unexpected start of line. If breaking lines, |/||/&& should be at the end of the previous one.
> 
> 
> In files/disable_inactive_freeipa_users.sh line [63](https://github.com/cisagov/ansible-role-freeipa-server/actions/runs/10306946528/job/28535130174?pr=73#step:21:64):
>   users=$(
>   ^-- SC1009 (info): The mentioned syntax error was in this variable assignment.
>         ^-- SC1073 (error): Couldn't parse this command expansion. Fix to allow more checks.
> 
> 
> In files/disable_inactive_freeipa_users.sh line [68](https://github.com/cisagov/ansible-role-freeipa-server/actions/runs/10306946528/job/28535130174?pr=73#step:21:69):
>       | sed --quiet 's/^\s*User login:\s*//p'
>       ^-- SC1072 (error): Expected end of $(..) expression. Fix any mentioned problems and try again.
>       ^-- SC1133 (error): Unexpected start of line. If breaking lines, |/||/&& should be at the end of the previous one.
> 
> For more information:
>   https://www.shellcheck.net/wiki/SC1133 -- Unexpected start of line. If brea...
>   https://www.shellcheck.net/wiki/SC10[72](https://github.com/cisagov/ansible-role-freeipa-server/actions/runs/10306946528/job/28535130174?pr=73#step:21:73) -- Expected end of $(..) expression....
>   https://www.shellcheck.net/wiki/SC10[73](https://github.com/cisagov/ansible-role-freeipa-server/actions/runs/10306946528/job/28535130174?pr=73#step:21:74) -- Couldn't parse this command expan...
> ```
> 
> As a result I have dropped that commit.

@jsf9k It appears to be an issue with the comment in the middle of the `$()` expression. If I remove the comment explaining `--quiet` and reformat per my suggestion linting passes without issue. Otherwise I believe it needs to be formatted as it is now (and `shfmt` actually reformats to this if I adjust the pipes to the end of the preceding line).

jsf9k commented:
Can you please commit the change you want?

mcdonnnj commented on README.md:
```diff
@@ -14,7 +14,8 @@ None.
 |----------|-------------|---------|----------|
-| freeipa_days_before_inactive | The number of days a user can go without logging in before his or her account is determined to be inactive and is disabled. | `45` | No |
+| freeipa_days_before_inactive | The number of days an account can go without being authenticated before it is determined to be inactive and is disabled. | `45` | No |
+| freeipa_disable_inactive_users | Whether or not to set up FreeIPA and create a systemd timer and service to disable inactive users. | `false` | No |
```
Minor complaint but I find the phrasing a bit ambiguous around the "set up FreeIPA" part. It would be nice to tie it more directly to the second half of the sentence by providing some mention of what "set up FreeIPA" entails.

jsf9k replied:
Please see commit 494db64.

mcdonnnj reviewed (approved):
@jsf9k My reply around the script formatting was more for informative purposes. I'm not stressed enough about it to integrate the comment into the preceding comment block. I just wanted to document why it ended up needing to be formatted that way and what would need to be done if one cared enough to change the formatting. Aside from that this LGTM with one item for your consideration.

jsf9k committed:
Describe more specifically the change being made to FreeIPA

We are configuring FreeIPA to enable tracking of the most recent
authentication time for each user.

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>
